### PR TITLE
cmd/snap-bootstrap: allow non-gpt in fallback mode to support rpi

### PR DESCRIPTION
**Pull request by Meulengracht**
See related issue on LP: https://bugs.launchpad.net/canonical-kernel-snaps/+bug/2114779

### CODE REVIEW

File: cmd/snap-bootstrap/cmd_scan_disk.go [hunk 1]
```diff
@@ -106,7 +106,41 @@ func isGpt(probe blkid.AbstractBlkidProbe) bool {
...
+	probe.EnableSuperblocks(true)
+	probe.SetSuperblockFlags(blkid.BLKID_SUBLKS_LABEL | blkid.BLKID_SUBLKS_UUID)
+
+	if err := probe.DoSafeprobe(); err != nil {
+		return p, err
+	}
+
+	if val, err := probe.LookupValue("LABEL"); err != nil {
+		return p, err
+	} else {
```
Comment (alfonsosanchezbeato) [inline cmd/snap-bootstrap/cmd_scan_disk.go]: The else is not necessary and is a bit confusing

File: cmd/snap-bootstrap/cmd_scan_disk.go [hunk 2]
```diff
@@ -106,7 +106,41 @@ func isGpt(probe blkid.AbstractBlkidProbe) bool {
...
+		p.Name = val
+	}
+
+	// UUID is not required as a part of fallback mode, in that sense
+	// we just rely on the FS label
+	if !fallbackMode {
+		if val, err := probe.LookupValue("UUID"); err != nil {
+			return p, err
+		} else {
```
Comment (alfonsosanchezbeato) [inline cmd/snap-bootstrap/cmd_scan_disk.go]: Same here: re else. I prefer to have val declared before than this tbh.

File: cmd/snap-bootstrap/cmd_scan_disk.go [hunk 3]
```diff
@@ -132,9 +168,12 @@ func probePartitions(node string) ([]Partition, error) {
 
 	ret := make([]Partition, 0)
 	for _, partition := range partitions.GetPartitions() {
-		label := partition.GetName()
-		uuid := partition.GetUUID()
-		ret = append(ret, Partition{label, uuid})
+		pnode := fmt.Sprintf("%sp%d", node, partition.GetPartNo())
+		if p, err := probePartition(pnode, fallbackMode); err != nil {
+			return ret, err
+		} else {
```
Comment (alfonsosanchezbeato) [inline cmd/snap-bootstrap/cmd_scan_disk.go:175]: Same about else

File: cmd/snap-bootstrap/cmd_scan_disk.go [hunk 4]
```diff
@@ -106,7 +106,41 @@ func isGpt(probe blkid.AbstractBlkidProbe) bool {
 	return pttype == "gpt"
 }
 
-func probePartitions(node string) ([]Partition, error) {
+func probePartition(node string, fallbackMode bool) (Partition, error) {
```
Comment (alfonsosanchezbeato) [inline cmd/snap-bootstrap/cmd_scan_disk.go]: Afaict this returns now filesystem label/uuid, so it is more `probeFilesystem` than `probePartition` maybe. Shouldn't this return the partition label/uuid if we have GPT, and filesystem for MBR? Otherwise this is not keeping the old behavior.

Comment (Meulengracht) [reply cmd/snap-bootstrap/cmd_scan_disk.go]: You are absolutely right - I should preserve the normal behavior for !fallbackMode, probably explains why I broke all the nested tests for UC24 :-)

File: cmd/snap-bootstrap/cmd_scan_disk.go (hunk 1 above)
Comment (Meulengracht) [reply cmd/snap-bootstrap/cmd_scan_disk.go]: Changed!

File: cmd/snap-bootstrap/cmd_scan_disk.go (hunk 2 above)
Comment (Meulengracht) [reply cmd/snap-bootstrap/cmd_scan_disk.go]: Changed here as well

File: cmd/snap-bootstrap/cmd_scan_disk.go (hunk 4 above)
Comment (Meulengracht) [reply cmd/snap-bootstrap/cmd_scan_disk.go]: Changed this now!

File: cmd/snap-bootstrap/cmd_scan_disk.go [hunk 5]
```diff
@@ -115,13 +145,15 @@ func probePartitions(node string) ([]Partition, error) {
...
 
 	if err := probe.DoSafeprobe(); err != nil {
 		return nil, err
 	}
 
-	if !isGpt(probe) {
+	// In fallback mode, which is non-UEFI, it may be
+	// MBR disk schema and not gpt, so we still want to
+	// discover filesystem labels.
+	if !fallbackMode && !isGpt(probe) {
```
Comment (alfonsosanchezbeato) [inline cmd/snap-bootstrap/cmd_scan_disk.go]: I think that this condition is never happening, as it is returning if a uuid for a booted partition was found (no fallback) and the partition is MBR, but the former can happen only for GPT partitions. So you could remove it.

File: cmd/snap-bootstrap/cmd_scan_disk.go [hunk 6]
```diff
@@ -132,9 +164,21 @@ func probePartitions(node string) ([]Partition, error) {
 
 	ret := make([]Partition, 0)
 	for _, partition := range partitions.GetPartitions() {
-		label := partition.GetName()
-		uuid := partition.GetUUID()
-		ret = append(ret, Partition{label, uuid})
+		name := partition.GetName()
+		if name == "" && fallbackMode {
```
Comment (alfonsosanchezbeato) [inline cmd/snap-bootstrap/cmd_scan_disk.go]: Maybe instead of using fallbackMode, you could do one or another thing depending on result of `isGpt()`. I think that we actually do not need that variable at all.

Thinking about this, the call to `probePartitions` could happen before the call to `bootFindPartitionUUIDForBootedKernelDisk`, as this is done always both if the uuid is found or not. Furthermore, at that point we might want to call `bootFindPartitionUUIDForBootedKernelDisk` only if this is GPT. But maybe this is more of a TODO.

**Review by alfonsosanchezbeato — COMMENTED** [review]
Thanks, couple more comments

File: cmd/snap-bootstrap/cmd_scan_disk.go (hunk 6 above)
Comment (valentindavid) [reply cmd/snap-bootstrap/cmd_scan_disk.go]: I would not call `GetName()` if it is not a gpt. Also I wonder if we can have an empty name in gpt. So using gpt has a condition would be better.

File: cmd/snap-bootstrap/cmd_scan_disk.go [hunk 7]
```diff
@@ -106,7 +106,37 @@ func isGpt(probe blkid.AbstractBlkidProbe) bool {
...
+	val, err := probe.LookupValue("LABEL")
+	if err != nil {
+		return p, err
+	}
+	p.Name = val
+
+	// UUID is not required as a part of fallback mode
+	val, err = probe.LookupValue("UUID")
+	if err == nil {
+		p.UUID = val
```
Comment (valentindavid) [inline cmd/snap-bootstrap/cmd_scan_disk.go]: Please, do not put the filesystem uuid in there. This is for the partition UUID. Do not mix them up or it will be very difficult to understand what is happening.

File: cmd/snap-bootstrap/cmd_scan_disk.go [hunk 8]
```diff
@@ -106,7 +106,37 @@ func isGpt(probe blkid.AbstractBlkidProbe) bool {
...
+
+	if err := probe.DoSafeprobe(); err != nil {
+		return p, err
+	}
+
+	val, err := probe.LookupValue("LABEL")
+	if err != nil {
+		return p, err
+	}
+	p.Name = val
```
Comment (valentindavid) [inline cmd/snap-bootstrap/cmd_scan_disk.go]: Also here, I would not mix partition name and filesystem label. Please use a different field or different structure.

File: cmd/snap-bootstrap/cmd_scan_disk.go (hunk 6 above)
Comment (Meulengracht) [reply cmd/snap-bootstrap/cmd_scan_disk.go]: Updated, thank you for this input. Please re-review

File: cmd/snap-bootstrap/cmd_scan_disk.go (hunk 7 above)
Comment (Meulengracht) [reply cmd/snap-bootstrap/cmd_scan_disk.go]: Good point, I think you are right it needs proper distinction.

File: cmd/snap-bootstrap/cmd_scan_disk.go (hunk 8 above)
Comment (Meulengracht) [reply cmd/snap-bootstrap/cmd_scan_disk.go]: Added a different field, this makes sense, I should not mix up the values here.

File: cmd/snap-bootstrap/cmd_scan_disk.go [hunk 9]
```diff
@@ -106,35 +112,66 @@ func isGpt(probe blkid.AbstractBlkidProbe) bool {
...
 	if err != nil {
 		return nil, err
 	}
 
 	ret := make([]Partition, 0)
 	for _, partition := range partitions.GetPartitions() {
-		label := partition.GetName()
-		uuid := partition.GetUUID()
-		ret = append(ret, Partition{label, uuid})
+		if !gpt {
```
Comment (alfonsosanchezbeato) [inline cmd/snap-bootstrap/cmd_scan_disk.go]: nitpick: I'd prefer to invert conditions and have the `if gpt` part first, to avoid the negation.

**Review by alfonsosanchezbeato — DISMISSED** [review]
LGTM, thanks for the changs. Small nickpick.

When testing in the RPi, we should make sure that all modes are fine (try recover, install, etc.)

File: cmd/snap-bootstrap/cmd_scan_disk.go (hunk 9 above)
Comment (Meulengracht) [reply cmd/snap-bootstrap/cmd_scan_disk.go]: Agreed! Updated

**Comment by Meulengracht** [comment]
> LGTM, thanks for the changs. Small nickpick.
> 
> When testing in the RPi, we should make sure that all modes are fine (try recover, install, etc.)

Thank you - I'm retesting for both install and refresh case

File: cmd/snap-bootstrap/cmd_scan_disk.go [hunk 10]
```diff
@@ -106,35 +112,66 @@ func isGpt(probe blkid.AbstractBlkidProbe) bool {
...
-		uuid := partition.GetUUID()
-		ret = append(ret, Partition{label, uuid})
+		if gpt {
+			ret = append(ret, Partition{
+				Name: partition.GetName(),
+				UUID: partition.GetUUID(),
+			})
+		} else {
+			// For MBR we have to probe the filesystem for details
+			pnode := fmt.Sprintf("%sp%d", node, partition.GetPartNo())
```
Comment (valentindavid) [inline cmd/snap-bootstrap/cmd_scan_disk.go]: Not sure `"%sp%s"` works every time...

Comment (valentindavid) [reply cmd/snap-bootstrap/cmd_scan_disk.go]: I am surprised that we cannot get superblock info of the partitions from the partition table device probe... I have to look into that.

Comment (valentindavid) [reply cmd/snap-bootstrap/cmd_scan_disk.go]: Could you see if you could just use the offset of the partition and do a probe at the offset instead on relying on a different device node? I feel uncomfortable a udev program to look at other device nodes that the one it is given.

Comment (Meulengracht) [reply cmd/snap-bootstrap/cmd_scan_disk.go]: Sure, let me test that - that would allow us to do it without relying on other device nodes, so that sounds cool if it works

**Review by valentindavid — APPROVED** [review]
Thank you it looks good.

I wonder the corner case where it detects a partition table not matching the block size of the device. It will be fun the day Rasberry PI tries to move to 4k. Well, let's see.
